pkg/schedule/schedulers/balance_keyrange.go

```diff
@@ -0,0 +1,488 @@
+			if !fromHasRegion {
+				continue
+			}
+			if toHasRegion, exist := toStore.RegionIDSet[r]; !exist || !toHasRegion {
+				// If toStore doesn't has this region, then create a move op.
+				o.Regions[r] = false
+				o.OriginalPeer = fromStore.OriginalPeer[r]
+				toStore.RegionIDSet[r] = true
+				fromStore.RegionIDSet[r] = false
+				break
+			}
+		}
+	}
+	return &o
+}
+
+func buildMigrationPlan(stores []*storeRegionSet) ([]int, []int, []*migrationOp, int) {
+	totalPeersCount := 0
+	if len(stores) == 0 {
+		log.Info("no stores for migration")
+		return []int{}, []int{}, []*migrationOp{}, 0
+	}
+	for _, store := range stores {
+		totalPeersCount += len(store.RegionIDSet)
+	}
+	avr := totalPeersCount / len(stores)
+	remainder := totalPeersCount % len(stores)
+	// sort TiFlash stores by region count in descending order
+	slices.SortStableFunc(stores, func(lhs, rhs *storeRegionSet) int {
+		return -cmp.Compare(len(lhs.RegionIDSet), len(rhs.RegionIDSet))
+	})
+	expectedCount := []int{}
+	for range remainder {
+		expectedCount = append(expectedCount, avr+1)
+	}
+	for i := remainder; i < len(stores); i++ {
+		expectedCount = append(expectedCount, avr)
+	}
+
+	senders := []int{}
+	receivers := []int{}
+	sendersVolume := []int{}
+	receiversVolume := []int{}
+	for i, store := range stores {
+		if len(store.RegionIDSet) < expectedCount[i] {
+			receivers = append(receivers, i)
+			receiversVolume = append(receiversVolume, expectedCount[i]-len(store.RegionIDSet))
+		}
+		if len(store.RegionIDSet) > expectedCount[i] {
+			senders = append(senders, i)
+			sendersVolume = append(sendersVolume, len(store.RegionIDSet)-expectedCount[i])
+		}
+	}
+
+	ops := []*migrationOp{}
+	movements := 0
+	for i, senderIndex := range senders {
+		fromStore := stores[senderIndex]
+		for {
+			if sendersVolume[i] <= 0 {
+				break
+			}
+			for j, receiverIndex := range receivers {
+				toStore := stores[receiverIndex]
+				if receiversVolume[j] > 0 {
+					n := sendersVolume[i]
+					if n > receiversVolume[j] {
+						n = receiversVolume[j]
+					}
+					receiversVolume[j] -= n
+					sendersVolume[i] -= n
+					op := pickRegions(n, fromStore, toStore)
+					movements += len(op.Regions)
+					ops = append(ops, op)
+				}
+			}
+		}
+	}
+
+	return senders, receivers, ops, movements
+}
+
+type operatorWrapper struct {
+	Operator  *operator.Operator `json:"operators"`
+	Region    *core.RegionInfo
+	FromStore uint64
+	Peer      *metapb.Peer
+}
+type migrationPlan struct {
+	ErrorCode  uint64             `json:"error_code"`
+	StartKey   []byte             `json:"start_key"`
+	EndKey     []byte             `json:"end_key"`
+	Ops        []*migrationOp     `json:"ops"`
+	Operators  []*operatorWrapper `json:"operators"`
+	Running    []*operatorWrapper `json:"running"`
+	TotalCount int                `json:"total"`
+	StartTime  time.Time
+}
+
+func computeCandidateStores(requiredLabels []*metapb.StoreLabel, stores []*metapb.Store, regions []*core.RegionInfo) []*storeRegionSet {
```

Comment:
We need to decide who can be a candidate for each region according to the constraint. Not all regions share the same stores.